bluesky/callbacks/core.py

```diff
@@ -862,3 +873,166 @@ def descriptor(self, doc):
     def stop(self, doc):
         self.livefit.stop(doc)
         # Intentionally override LivePlot.stop. Do not call super().
+
+
+
+
+class SmartDispatcher(CallbackBase):
+    """
+    A callback that tries to guess useful callbacks and fan out documents.
+
+    .. attribute:: muted
+        Set to ``True`` to temporarily turn this off.
+
+    .. attribute:: x
+        Field name (such as 'theta') or ``ExogenousVars.seq_num`` or
+        ``ExogenousVars.time``
+
+    .. attribute:: ys
+        List of field names to plot against x
+
+    .. attribute:: overplot
+        Set to ``False`` to generate a new figure even when axes labels are
+        identical to those on a figure that is already open.
+
+    .. attribute :: cols
+        List of field names to include in table. (Event time and sequence
+        number are always included.)
+    """
+    def __init__(self):
+        self.active_cbs = []
+        self.muted = False
+
+        # plot settings
+        self.ys = []
+        self.x = None
+        self.overplot = False
+
+        # table settings
+        self.cols = []
+
+        # peak stats
+        self.ps = None
+        self.ps_field = None
+        self.ps_config = None
+
+        self.__sniff_dim = None
+        self.__start_doc = None
+
+    def _figure_name(self, base_name):
+        """Helper to compute figure name
+
+        This takes in a base name an return the name of the figure to use.
+
+        If self.overplot, then this is a no-op.
+        If not self.overplot then append '(N)' to the end of the string until a
+        non-existing figure is found
+        """
+        if not self.overplot:
+            if not plt.fignum_exists(base_name):
+                pass
+            else:
+                for j in itertools.count(1):
+                    numbered_template = '{} ({})'.format(base_name, j)
+                    if not plt.fignum_exists(numbered_template):
+                        base_name = numbered_template
+                        break
+        return base_name
+
+    def clean(self):
+        self.active_cbs.clear()
+        self.__sniff_dim = None
+        self.__start_doc = None
+        self.ps = None
+
+    def start(self, doc):
+        # Inspect the start document and try to guess what some useful
+        # callbacks will be.
+
+        self.clean()
+        self.__start_doc = doc
+        if self.muted:
+            return
+
+        # We can always set up some table.
+        motors = doc.get('motors', [])
+        table = LiveTable(list(motors) + list(self.ys) + list(self.cols))
```

Comment:
does this assume that the motors names are the same as their primary readout?  I do not think this was an issue before because we had access to the actual motor objects.